File: app/src/main/java/com/firebase/mark42/databasedemo/java/UserViewModel.java
package com.firebase.mark42.databasedemo.java;

import android.os.AsyncTask;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModel;

import com.firebase.mark42.databaseassistant.DatabaseResult;
import com.firebase.mark42.databasedemo.DATABASE_KEY;
import com.firebase.mark42.databasedemo.R;
import com.firebase.mark42.databasedemo.User;

import java.net.URL;
import java.util.concurrent.ExecutionException;

public class UserViewModel extends ViewModel {

    //private UserRepo userRepo = new UserRepo();
    private UserRepo userRepo = UserRepo.getInstance();

    public void getUserFromDatabase(AppCompatActivity activity) {
        final TextView fName = activity.findViewById(R.id.firstName);
        final TextView lName = activity.findViewById(R.id.lastName);
        final TextView email = activity.findViewById(R.id.email);

        userRepo.getFromDatabase("users/-LxjJXsJdbZiEjjh1A89").observe(activity, new Observer<User>() {
            @Override
            public void onChanged(User user) {
                fName.setText(user.getFirstName());
                lName.setText(user.getLastName());
                email.setText(user.getEmail());
            }
        });
    }

    public void getUserFromCache(final AppCompatActivity activity) {
        getUserFromCache().observe(activity, new Observer<User>() {
            @Override
            public void onChanged(User user) {
                TextView fName = activity.findViewById(R.id.firstName);
                TextView lName = activity.findViewById(R.id.lastName);
                TextView email = activity.findViewById(R.id.email);

                fName.setText(user.getFirstName());
                lName.setText(user.getLastName());
                email.setText(user.getEmail());
            }
        });
    }

    private static LiveData<User> getUserFromCache() {
        final UserRepo userRepo = UserRepo.getInstance();
        final MutableLiveData<User> liveData = new MutableLiveData<>();
        new AsyncTask<Void, Void, LiveData<User>>() {

            @Override
            protected LiveData<User> doInBackground(Void... voids) {
                try {
                    User result = userRepo.getFromDatabaseCache
                            ("users/-LxjJXsJdbZiEjjh1A89").get();
                    liveData.postValue(result);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                } catch (ExecutionException e) {
                    e.printStackTrace();
                }
                return liveData;
            }
        }.execute();

        return liveData;
    }

    public static LiveData<DatabaseResult> deleteFromDatabase() {
        final UserRepo userRepo = UserRepo.getInstance();
        final MutableLiveData<DatabaseResult> liveData = new MutableLiveData<>();
        new AsyncTask<Void, Void, LiveData<DatabaseResult>>() {

            @Override
            protected LiveData<DatabaseResult> doInBackground(Void... voids) {
                try {
                    DatabaseResult result = userRepo.deleteFromDatabase("").get();
                    liveData.postValue(result);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                } catch (ExecutionException e) {
                    e.printStackTrace();
                }
                return liveData;
            }
        }.execute();
        return liveData;
    }
}

File: app/src/main/java/com/firebase/mark42/databasedemo/UserViewModel.kt
package com.firebase.mark42.databasedemo

import android.content.Context
import android.util.Log
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.launch
import java.util.HashMap

class UserViewModelFactory(private val context: Context) : ViewModelProvider.Factory {
    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        return UserViewModel(context) as T
    }
}

class UserViewModel(context: Context) : ViewModel() {

    private val userRepo by lazy {
        UserRepo.getInstance(context)
    }

    fun getUserFromDatabase(activity: AppCompatActivity) {
        val fName = activity.findViewById<TextView>(R.id.firstName)
        val lName = activity.findViewById<TextView>(R.id.lastName)
        val email = activity.findViewById<TextView>(R.id.email)
        userRepo.getFromDatabase(UserRepo.userPath("-LxjJXsJdbZiEjjh1A89")).observe(activity, Observer { user ->
            fName.text = user?.firstName
            lName.text = user?.lastName
            email.text = user?.email
        })
    }

    fun getUserFromCache(activity: AppCompatActivity) {
        viewModelScope.launch {
            val fName = activity.findViewById<TextView>(R.id.firstName)
            val lName = activity.findViewById<TextView>(R.id.lastName)
            val email = activity.findViewById<TextView>(R.id.email)
            val user = userRepo.getFromDatabaseCache(UserRepo.userPath("-LxjJXsJdbZiEjjh1A89"))
            fName.text = user?.firstName
            lName.text = user?.lastName
            email.text = user?.email
        }
    }

    fun pushUserToDatabase(user: User) {
        viewModelScope.launch {
            //it return the firebase push id key
            val key = userRepo.pushToDatabase(UserRepo.path(), user)
            user.userPushId = key
            // show confirmation message to user
        }
    }

    fun updateUserFieldToDatabase() {
        viewModelScope.launch {
            val result = userRepo.updateChildToDatabase(UserRepo.userPath(
                "-LxjJoZry_g7uqwP3RRZ/" + DATABASE_KEY.USER.firstName), "Sachin")
            if (result.isSuccess()) {
                //Display success message
            } else {
                //Display error message
                result.errorMessage
            }
        }
    }

    fun updateUserToDatabase(updates: HashMap<String, Any?>) {
        viewModelScope.launch {
            //if path is null then it'll create a new node
            val result = userRepo.updateChildrenToDatabase(UserRepo.userPath(
                "-LzhYTCGOOYIQt1uOq7B"), updates)
            if (result.isSuccess()) {
                //Display success message
            } else {
                //Display error message
                result.errorMessage
            }
        }
    }

    fun deleteFromDatabase() {
        viewModelScope.launch {
            val result = userRepo.deleteFromDatabase(UserRepo.userPath(
                "-LzhYTCGOOYIQt1uOq7B/" + DATABASE_KEY.USER.email))
            if (result.isSuccess()) {
                //Display success message
            } else {
                //Display error message
                result.errorMessage
            }
        }
    }
}